trees/trees.go

```diff
@@ -34,6 +34,86 @@ import (
 
 type treeKey struct{}
 
+type accessRule struct {
+	// Tree states are accepted if there is a 'true' value for them in this map.
+	okStates map[trillian.TreeState]bool
+	// Allows the error code to be specified for specific rejected states.
+	rejectCodes map[trillian.TreeState]codes.Code
+	// Tree types are accepted if there is a 'true' value for them in this map.
+	okTypes map[trillian.TreeType]bool
+	// okWrite is true if write access should be allowed.
+	okWrite bool
+}
+
+// These rules define the permissible combinations of tree state and type
+// for each operation type.
+var rules = map[OpType]accessRule{
+	Unknown: {},
+	Admin: {
+		okStates: map[trillian.TreeState]bool{
+			trillian.TreeState_UNKNOWN_TREE_STATE: true,
+			trillian.TreeState_ACTIVE:             true,
+			trillian.TreeState_DRAINING:           true,
+			trillian.TreeState_FROZEN:             true,
+		},
+		okTypes: map[trillian.TreeType]bool{
+			trillian.TreeType_LOG:            true,
+			trillian.TreeType_MAP:            true,
+			trillian.TreeType_PREORDERED_LOG: true,
+		},
+		okWrite: true,
+	},
+	Query: {
+		okStates: map[trillian.TreeState]bool{
+			// Have to allow queries on unknown state so storage can get a chance
+			// to return ErrTreeNeedsInit.
+			trillian.TreeState_UNKNOWN_TREE_STATE: true,
+			trillian.TreeState_ACTIVE:             true,
+			trillian.TreeState_DRAINING:           true,
+			trillian.TreeState_FROZEN:             true,
+		},
+		okTypes: map[trillian.TreeType]bool{
+			trillian.TreeType_LOG:            true,
+			trillian.TreeType_MAP:            true,
+			trillian.TreeType_PREORDERED_LOG: true,
+		},
+		okWrite: false,
+	},
+	QueueLog: {
+		okStates: map[trillian.TreeState]bool{
+			trillian.TreeState_ACTIVE: true,
+		},
+		rejectCodes: map[trillian.TreeState]codes.Code{
+			trillian.TreeState_DRAINING: codes.PermissionDenied,
+		},
+		okTypes: map[trillian.TreeType]bool{
+			trillian.TreeType_LOG:            true,
+			trillian.TreeType_PREORDERED_LOG: true,
+		},
+		okWrite: true,
+	},
+	SequenceLog: {
+		okStates: map[trillian.TreeState]bool{
+			trillian.TreeState_ACTIVE:   true,
+			trillian.TreeState_DRAINING: true,
```

Comment:
Does `DRAINING` make sense for `PREORDERED_LOG`? There is no queue in that mode.

However, I can imagine `DRAINING` mean "still has sequenced entries not integrated into the tree, hence the sequencer is still working". The caveat though is that there might be gaps in sequenced entries, so in this sense the `DRAINING` state will last forever.